### PR TITLE
feat(payments-plugin): Use Mollie's Order API

### PR DESCRIPTION
**Pull request by martijnvdbrug**
See #1824 for more details.

This PR implements the Mollie orders API, so we can use pay later methods.

mollie

Important changes:
* Orders MUST have a shipping or billing address with at least city, postalCode, street and country. Payment intent creation will fail if this is not the case
* You can now use pay-later methods: Orders with a pay-later method will be in PaymentAuthorized and need to be settled/captured via the admin UI or programatically. Mollie wants you to do this when you actually ship the order. The payment MUST be captured within 28 days, otherwise the order in Mollie will expire, and your payment can never be captured/settled. https://docs.mollie.com/orders/status-changes#expired

**AutoCapture**
If you don't want to worry about capture/settlement, you can configure auto-capture per payment method. Auto-capture immediately captures/settles the payment after being authorized. This can be useful for digital products or services, where there is no physical shipping moment.

Inspired by the way Shopify handles pay-later methods with auto capturing: https://docs.klarna.com/platform-solutions/shopify/payments/klarna-payments/order-management/

New e2e tests added:
image

### CODE REVIEW

File: packages/payments-plugin/README.md [hunk 1]
```diff
@@ -1,3 +1,19 @@
 # Payments plugin
 
 For documentation, see https://www.vendure.io/docs/typescript-api/payments-plugin
+
+## Development
+
+### Mollie local development
```
Comment (martijnvdbrug) [inline packages/payments-plugin/README.md:7]: Move dev docs to readme instead of public plugin docs, because its only for contributors.

File: packages/payments-plugin/e2e/mollie-dev-server.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import { molliePaymentHandler } from '../package/mollie/mollie.handler';
+
+import { CREATE_PAYMENT_METHOD } from './graphql/admin-queries';
+import { CreatePaymentMethod } from './graphql/generated-admin-types';
+import { AddItemToOrder } from './graphql/generated-shop-types';
+import { ADD_ITEM_TO_ORDER } from './graphql/shop-queries';
+import { CREATE_MOLLIE_PAYMENT_INTENT, setShipping } from './payment-helpers';
+
+/**
+ * This should only be used to locally test the Mollie payment plugin
```
Comment (martijnvdbrug) [inline packages/payments-plugin/e2e/mollie-dev-server.ts:28]: Do we want this file in GIT? This is meant for contributors (as mentioned in the readme). It's quite the hassle to setup a local env that actually communicates with the live Mollie env, so this file is very useful for any future contributors (and me for any future changes)

This dev-server only requires a Mollie apiKey and creates ready to test orders.

File: packages/payments-plugin/src/mollie/mollie.helpers.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            givenName: customer.firstName,
+            familyName: customer.lastName,
+            email: customer.emailAddress,
+        };
+    }
+}
+
+/**
+ * Map all order and shipping lines to a single array of Mollie order lines
+ */
```
Comment (martijnvdbrug) [inline packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: @michaelbromley Will he function below always result in the order total? Mollie will validate if all lines add up to the order total, including taxes and discounts. If not, Mollie says no, and we won't be able to create a payment intent.

I don't like recreating this logic, but I've seen this a few times now: accounting software and external fulfillment platforms that want all lines with prices, taxes, shipping and discounts AND want it to add up to the order total.

**Comment by martijnvdbrug** [comment]
@michaelbromley Build fails with `Error: ../core/src/i18n/i18n.service.ts(7,21): error TS7016: Could not find a declaration file for module 'i18next-fs-backend'. '/home/runner/work/vendure/vendure/node_modules/i18next-fs-backend/cjs/index.js`
But I'm pretty sure this is not related to this PR?

File: packages/payments-plugin/README.md (hunk 1 above)
Comment (michaelbromley) [reply packages/payments-plugin/README.md:7]: Yes this makes sense.

File: packages/payments-plugin/e2e/mollie-dev-server.ts (hunk 2 above)
Comment (michaelbromley) [reply packages/payments-plugin/e2e/mollie-dev-server.ts:28]: Yes for sure, keep this in the repo!

File: packages/payments-plugin/package.json [hunk 4]
```diff
@@ -1,40 +1,41 @@
...
-        "@vendure/common": "^1.8.4",
-        "@vendure/core": "^1.8.4",
-        "@vendure/testing": "^1.8.4",
-        "braintree": "^3.0.0",
-        "nock": "^13.1.4",
-        "rimraf": "^3.0.2",
-        "stripe": "^8.197.0",
-        "typescript": "4.3.5"
-    }
+  "name": "@vendure/payments-plugin",
```
Comment (michaelbromley) [inline packages/payments-plugin/package.json]: Looks like some unnecessary formatting of indentation happened here.

File: packages/payments-plugin/src/mollie/mollie.helpers.ts (hunk 3 above)
Comment (michaelbromley) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: The logic looks correct for the order lines themselves. The order total is further composed of the shipping and any surcharges. You can see the entire internal calculation for totals here: https://github.com/vendure-ecommerce/vendure/blob/81426fa01ad1ff82a587444c3768944e99c5e6b6/packages/core/src/service/helpers/order-calculator/order-calculator.ts#L453-L478

File: packages/payments-plugin/src/mollie/mollie.helpers.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+export function toAmount(value: number, orderCurrency: string): Amount {
+    return {
+        value: (value / 100).toFixed(2),
+        currency: orderCurrency,
+    };
+}
+
+/**
+ * Lookup one of Mollies allowed locales based on an orders countrycode or channel default.
+ * If both lookups fail, resolve to en_US to prevent payment failure
```
Comment (michaelbromley) [inline packages/payments-plugin/src/mollie/mollie.helpers.ts:72]: What does `locale` signify in Mollie? Is it only for localization? If so this is fine.

File: packages/payments-plugin/src/mollie/mollie.service.ts [hunk 6]
```diff
@@ -68,16 +75,16 @@ export class MollieService {
...
         const [order, paymentMethod] = await Promise.all([
-            this.activeOrderService.getOrderFromContext(ctx),
+            this.activeOrderService.getActiveOrder(ctx, undefined),
             this.getPaymentMethod(ctx, paymentMethodCode),
         ]);
         if (!order) {
             return new PaymentIntentError('No active order found for session');
         }
-        await this.entityHydrator.hydrate(ctx, order, { relations: ['lines', 'customer', 'shippingLines'] });
+        await this.entityHydrator.hydrate(ctx, order, { relations: ['customer', 'lines', 'lines.productVariant', 'shippingLines', 'shippingLines.shippingMethod'] });
```
Comment (michaelbromley) [inline packages/payments-plugin/src/mollie/mollie.service.ts]: With EntityHydrator, you don't need to list the intermediate joins like you do with TypeORM APIs. I.e you can put `lines.productVariant` and it will automatically know to join `lines` first.

**Review by michaelbromley — COMMENTED** [review]
Martijn, excellent PR! I just left a few comments, nothing major to change as far as I can see.

File: packages/payments-plugin/src/mollie/mollie.helpers.ts (hunk 3 above)
Comment (martijnvdbrug) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: I have included shipping in the calculation, but surcharges...

Are surcharges included in the `line.proratedLinePrice` price? 

Mollie only knows orderLines, it doesn't know the concept of surcharges. The docs state this:

> The total amount of the line, including VAT and discounts. Adding all totalAmount values together should result in the same amount as the amount top level property.

So, the sum of `lines.totalAmount` must equal the `order.totalAmount`. 

Is the solution here to add surcharges as separate lines?

Comment (michaelbromley) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: > Are surcharges included in the line.proratedLinePrice price?

No, they are separate.

In that case then yes, map surcharges to lines and Mollie should be happy.

File: packages/payments-plugin/src/mollie/mollie.helpers.ts (hunk 5 above)
Comment (martijnvdbrug) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:72]: Yep, only for showing the customers language during the hosted checkout. It is a mandatory field for some reason.

> Allows you to preset the language to be used in the hosted payment pages shown to the consumer.

**Comment by martijnvdbrug** [comment]
> > But I'm pretty sure this is not related to this PR. Should I try and fix it in this PR?
> 
> No ignore it for now. If you wanna have a go at fixing, do it in a separate PR.

It does also block the e2e tests right now, is that an issue?

**Comment by michaelbromley** [comment]
> It does also block the e2e tests right now, is that an issue?

Ah ok yeah we should fix that then. I'll take a look and see if I can figure out what's causing it...

File: packages/payments-plugin/src/mollie/mollie.handler.ts [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
     CreatePaymentErrorResult,
     CreatePaymentResult,
     CreateRefundResult,
     Logger,
     PaymentMethodHandler,
-    PaymentMethodService,
     SettlePaymentResult,
 } from '@vendure/core';
-import { Permission } from '@vendure/core';
+import { PaymentState } from '@vendure/core/src';
```
Comment (michaelbromley) [inline packages/payments-plugin/src/mollie/mollie.handler.ts]: This is the cause of the build error. Should be imported from @vendure/core.

File: packages/payments-plugin/src/mollie/mollie.service.ts [hunk 8]
```diff
@@ -1,9 +1,15 @@
...
+import { CreateParameters } from '@mollie/api-client/dist/types/src/binders/orders/parameters';
 import { Inject, Injectable } from '@nestjs/common';
 import {
     ActiveOrderService,
     ChannelService,
     EntityHydrator,
+    ErrorResult,
     LanguageCode,
     Logger,
     Order,
```
Comment (michaelbromley) [inline packages/payments-plugin/src/mollie/mollie.service.ts:15]: Add the `OrderStateTransitionError` import here rather than the current deep import.

**Review by michaelbromley — CHANGES_REQUESTED** [review]
Found the build issue

File: packages/payments-plugin/src/mollie/mollie.handler.ts (hunk 7 above)
Comment (martijnvdbrug) [reply packages/payments-plugin/src/mollie/mollie.handler.ts]: Oops, so it IS caused by this PR :1st_place_medal:

File: packages/payments-plugin/src/mollie/mollie.helpers.ts (hunk 3 above)
Comment (martijnvdbrug) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: @michaelbromley Not really familiar with surcharges. How can I add a promotion that will add a surcharge to the order? I would like to test if Mollie allows negative numbers on an order line + include a surcharge in the e2e test.

Comment (michaelbromley) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: Surcharges are not related to Promotions. Promotions can only make "adjustments", which is what we already take into account in the `discountedPrice` and `proratedPrice`. However, promotions _can_ make an orderline negative - this is an edge case and almost always an error in the PromotionAction logic. But should be gracefully handled if it comes up.

Surcharges must be manually added using `OrderService.addSurchargeToOrder`. There is currently no built-in mutation for this.

Comment (martijnvdbrug) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: Thanks, I will try to fix this today.

Comment (martijnvdbrug) [reply packages/payments-plugin/src/mollie/mollie.helpers.ts:28]: image
Mollie says 'OK'. I will add it to the e2e test and round up this PR

**Comment by michaelbromley** [comment]
Martijn, thanks so much for the huge amount of work you put into this! 🙏 🥳

**Comment by martijnvdbrug** [comment]
Happy to contribute! (And a healthy dose of self-interest, of course)
